src/server/game/Entities/Unit/Unit.cpp

```diff
@@ -12647,14 +12647,19 @@ bool Unit::IsFalling() const
 
 bool Unit::CanSwim() const
 {
+    if (IsPet())
+        return true;
+    if (!CanEnterWater())
+        return false;
+
     // Mirror client behavior, if this method returns false then client will not use swimming animation and for players will apply gravity as if there was no water
     if (HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_CANNOT_SWIM))
         return false;
     if (HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_PLAYER_CONTROLLED)) // is player
         return true;
     if (HasFlag(UNIT_FIELD_FLAGS_2, 0x1000000))
         return false;
-    if (IsPet() && HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_PET_IN_COMBAT))
+    if (HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_PET_IN_COMBAT))
         return true;
     return HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_RENAME | UNIT_FLAG_SWIMMING);
```

Comment:
why is UNIT_FLAG_RENAME here xD? What does that have to do with swimming?

Reply:
I know it seems stupid but that function was ripped from the client and that was the flag value used there